### ISSUE TITLE
Pass context to parse methods, to use it in refinements

### ISSUE THREAD

**Issue by julianomqs** [vriad/zod#148]
Hello,

I'm using Zod in the backend, and GraphQL.

I have the following code:

```typescript
// zod schema
const CREATE_SCHEMA = z.object({
  cliente: z.number()
  // other properties...
});

// GraphQL resolver
const resolver: Resolvers = {
  Mutation: {
    createChequeDevolvido: (parent, { input }, context) => {
      CREATE_SCHEMA.safeParse(input);
      // persist validated input in database...
    }
  }
};
```

The `cliente` property is an id to the table `cliente`, and I want to validate if it exists.

So it would look like this (using Objection.js):

```typescript
// zod schema
const CREATE_SCHEMA = z.object({
  cliente: z.number().refine(async arg => Cliente.query().where("id", arg).first(), {
    message: "Cliente não existe no banco de dados."
  });
  // other properties...
});
```

But I need to add the `tenant` key to that query, but the `tenant` key is only avaliable in the `context` argument within the GraphQL resolver.

In Joi, you can pass a `context` property in the `validate` methods. This `context`is a bag of values to be used in the validation schemas.

By having something like that in Zod, we pass a `context` option in the `parse` methods, and this `context`is avaliable as a parameter in the `refine`method, and other places if necessary.

I think that would solve my problem. Something like this:

```typescript
// zod schema
const CREATE_SCHEMA = z.object({
  cliente: z.number().refine(async (arg, context) => Cliente.query()
      .where("id", arg)
      .where("tenant", context.tenant)
      .first(), {
    message: "Cliente não existe no banco de dados."
  });
  // other properties...
});

// GraphQL resolver
const resolver: Resolvers = {
  Mutation: {
    createChequeDevolvido: (parent, { input }, context) => {
      CREATE_SCHEMA.safeParse(input, {
        context: {
          tenant: context.tenant // the context here is the GraphQL context
        }
      });
      // persist validated input in database...
    }
  }
};
```

In short, this would add the possibility of using external, runtime values in validations.

What do you think?

Sorry if this was in docs or in another issue, but I couldn't find it.

Thanks.

**Comment by vriad** [comment]
This does exist elsewhere: https://github.com/vriad/zod/issues/84 - that issue also contains some of my reasons for disliking contextual validations in general. I consider it a huge antipattern and prefer that people declare all their schemas explicitly. In fact I'm opposed to "parse parameters" in general — I think the schema parsing should be a "pure function" it that it takes a single argument and returns a single output in a deterministic way. 

Though the bigger problem with your use case is that async refinements don't work at all currently - they're coming in v2.

In this case I would strongly recommend just creating a modified version of your schema within the resolver scope. It's far more explicit and easier to understand the working of the code.

```ts
// define base schema
const CREATE_SCHEMA_BASE = z.object({
  cliente: z.number()
  // other properties...
});

// in resolver
const cliente = await Cliente.query()
      .where("id", arg)
      .where("tenant", context.tenant)
      .first();
const CREATE_SCHEMA = CREATE_SCHEMA_BASE.extend({
  cliente: CREATE_SCHEMA_BASE.shape.cliente.refine(() => cliente, {
    message: "Cliente não existe no banco de dados."
  });
})
``

**Comment by julianomqs** [comment]
Yeah, I agree contextual validation would be harder to understand.

Your solution works but I have to verify before each query if the `id` was sent, so it is not the ideal. The async validations implementation would solve that issue.

Since there are issue for contextual validation and async validation, I'm closing this issue.

Thanks.

**Comment by IlyaSemenov** [comment]
I think a good example demonstrating the need of context would be parsing data in the context of the database transaction:

```ts
const user_schema = z.object({
  username: z.string().refine(
    // Imaginary code: pull trx (database transaction) from context.
    // This is important to prevent race condition.
    async (username, { trx }) =>
      !(await UserModel.query(trx).where({ username }).first()),
    {
      message: 'This username is already taken.',
    },
  ),
})

router.post('/register', async ctx => {
  await UserModel.transaction(async trx => {
    // Imaginary code: provide trx (database transaction) as a context.
    const data = await user_schema.parseAsync(ctx.request.body, { trx })
    const user = await UserModel.query(trx).insert(data)
    ctx.body = { user }
  })
})
```

Alas, that doesn't work.

The approach suggested above (doing extends accessing the inner shapes) leads to unmaintainable code and is hard to reuse across different invocations. The reasonable approach would be create ad-hoc schemas in runtime in the lexical context (closure) with the desired data. Something like:

```ts
const user_schema = (trx: Transaction) =>
  z.object({
    username: z
      .string()
      .refine(
        async username =>
          !(await UserModel.query(trx).where({ username }).first()),
        {
          message: 'This username is already taken.',
        },
      ),
  })

router.post('/register', async ctx => {
  await UserModel.transaction(async trx => {
    const data = await user_schema(trx).parseAsync(ctx.request.body)
    const user = await UserModel.query(trx).insert(data)
    ctx.body = { user }
  })
})
```

I wonder how does that affect performance though...

**Comment by strblr** [comment]
Another really useful usecase would be using dataloaders in async transforms.

**Comment by IlyaSemenov** [comment]
Just a heads up, one can pass context into the refine callback via [AsyncLocalStorage](https://nodejs.org/api/async_context.html#asynclocalstoragerunstore-callback-args).
